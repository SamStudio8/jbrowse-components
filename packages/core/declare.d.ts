declare module 'shortid'
declare module 'object.fromentries'
