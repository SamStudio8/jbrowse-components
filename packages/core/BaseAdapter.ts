import { Observer, Observable } from 'rxjs'
import { ObservableCreate } from './util/rxjs'
// @ts-ignore
import { checkAbortSignal } from './util'

export interface Region {
  refName: string
  start: number
  end: number
}
interface BaseOptions {
  signal: AbortSignal
}
/**
 * Base class for adapters to extend. Defines some methods that subclasses must
 * implement.
 */
export default class BaseAdapter {
  // List of all possible capabilities. Don't un-comment them here.
  static capabilities = [
    // 'getFeatures',
    // 'getRefNames',
    // 'getRegions',
    // 'getRefNameAliases',
  ]

  constructor(config: any) {
    if (new.target === BaseAdapter) {
      throw new TypeError(
        'Cannot create BaseAdapter instances directly, use a subclass',
      )
    }
  }

  /**
   * Subclasses should override this method. Method signature here for reference.
   * @returns {Promise<string[]>} Array of reference sequence names used by the
   * source being adapted.
   */
  async getRefNames():Promise<string[]> {
    throw new Error('getRefNames should be overridden by the subclass')
    // Subclass method should look something like this:
    // await this.setup()
    // const { refNames } = this.metadata
    // return refNames
  }

  /**
   * Subclasses should override this method. Method signature here for reference.
   * @param {Region} region
   * @param {BaseOptions} options
   * @returns {Observable[Feature]} Observable of Feature objects in the region
   */
  // eslint-disable-next-line no-unused-vars
  getFeatures({ refName, start, end }: Region, opts: BaseOptions):Observable<any> {
    throw new Error('getFeatures should be overridden by the subclass')
    // Subclass method should look something like this:
    // return ObservableCreate(observer => {
    //   const records = getRecords(assembly, refName, start, end)
    //   records.forEach(record => {
    //     observer.next(this.recordToFeature(record))
    //   })
    //   observer.complete()
    // })
  }

  /**
   * Subclasses should override this method. Method signature here for reference.
   *
   * called to provide a hint that data tied to a certain region
   * will not be needed for the forseeable future and can be purged
   * from caches, etc
   * @param {Region} region
   */
  // eslint-disable-next-line no-unused-vars
  freeResources(region: Region) {
    throw new Error('freeResources should be overridden by the subclass')
  }

  /**
   * Checks if the store has data for the given assembly and reference
   * sequence, and then gets the features in the region if it does.
   * @param {Region} region see getFeatures()
   * @param {AbortSignal} [signal] optional AbortSignal for aborting the request
   * @returns {Observable[Feature]} see getFeatures()
   */
  getFeaturesInRegion(region: Region, opts: BaseOptions):Observable<any> {
    return ObservableCreate(async (observer:Observer<any>) => {
      const hasData = await this.hasDataForRefName(region.refName)
      checkAbortSignal(opts.signal)
      if (!hasData) {
        observer.complete()
      } else {
        this.getFeatures(region, opts).subscribe(observer)
      }
    })
  }

  /**
   * Check if the store has data for the given reference name. Also checks the
   * assembly name if one was specified in the initial adapter config.
   * @param {string} refName Name of the reference sequence
   * @returns {Promise<boolean>} Whether data source has data for the given
   * reference name
   */
  async hasDataForRefName(refName:string):Promise<boolean> {
    const refNames = await this.getRefNames()
    if (refNames.includes(refName)) return true
    return false
  }
}
