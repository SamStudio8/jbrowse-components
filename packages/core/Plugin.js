export default class Plugin {
  install() {
    this.installed = true
  }

  configure() {
    this.configured = true
  }
}
