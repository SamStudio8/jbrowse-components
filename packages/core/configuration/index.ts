export {
  ConfigurationSchema,
  ConfigurationSchemaType,
  ConfigurationSchemaDefinition,
  ConfigurationModel,
  ConfigurationReference,
  isConfigurationModel,
  isConfigurationSchemaType,
  isBareConfigurationSchemaType,
} from './configurationSchema'

export { getConf, readConfObject } from './util'
