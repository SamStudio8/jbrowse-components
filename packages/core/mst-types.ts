import shortid from 'shortid'
import { types, SnapshotOut } from 'mobx-state-tree'
import propTypes from 'prop-types'
import { PropTypes as MxPropTypes } from 'mobx-react'

import { assembleLocString } from './util'

export const ElementId = types.optional(types.identifier, shortid.generate)

// PropTypes that are useful when working with instances of these in react components
export const PropTypes = {
  Region: propTypes.shape({
    refName: propTypes.string.isRequired,
    start: propTypes.number.isRequired,
    end: propTypes.number.isRequired,
  }),
  ConfigSchema: MxPropTypes.objectOrObservableObject,
  Feature: propTypes.shape({
    get: propTypes.func.isRequired,
    id: propTypes.func.isRequired,
  }),
}

export const Region = types
  .model('Region', {
    refName: types.string,
    start: types.number,
    end: types.number,
  })
  .views(self => ({
    get locString() {
      return assembleLocString(self)
    },
  }))

export type IRegion = SnapshotOut<typeof Region>

export const FileLocalPath = types.model('FileLocalPath', {
  localPath: types.string, // TODO: refine
})

export const Uri = types.model('Uri', {
  uri: types.string, // TODO: refine
})

export const FileLocation = types.union(FileLocalPath, Uri)
