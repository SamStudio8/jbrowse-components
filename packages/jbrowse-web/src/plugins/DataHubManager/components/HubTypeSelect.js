import FormControl from '@material-ui/core/FormControl'
import FormControlLabel from '@material-ui/core/FormControlLabel'
import FormHelperText from '@material-ui/core/FormHelperText'
import Radio from '@material-ui/core/Radio'
import RadioGroup from '@material-ui/core/RadioGroup'
import propTypes from 'prop-types'
import React from 'react'

const hubTypeDescriptions = {
  ucsc: (
    <FormHelperText>
      A track or assembly hub in the{' '}
      <a
        href="http://genome.ucsc.edu/goldenPath/help/hgTrackHubHelp.html#Intro"
        rel="noopener noreferrer"
        target="_blank"
      >
        Track Hub
      </a>{' '}
      format
    </FormHelperText>
  ),
  jbrowse1: (
    <FormHelperText>
      A{' '}
      <a href="https://jbrowse.org/" rel="noopener noreferrer" target="_blank">
        JBrowse 1
      </a>{' '}
      data directory
    </FormHelperText>
  ),
}

function HubTypeSelect(props) {
  const { hubType, setHubType, enableNext } = props
  const hubTypes = [
    { value: 'ucsc', label: 'Track or Assembly Hub' },
    { value: 'jbrowse1', label: 'JBrowse Hub' },
  ]
  return (
    <FormControl component="fieldset">
      <RadioGroup
        value={hubType}
        onChange={event => {
          setHubType(event)
          enableNext()
        }}
      >
        {hubTypes.map(entry => (
          <FormControlLabel
            key={entry.value}
            control={<Radio />}
            value={entry.value}
            label={entry.label}
          />
        ))}
      </RadioGroup>
      {hubTypeDescriptions[hubType]}
    </FormControl>
  )
}

HubTypeSelect.defaultProps = {
  hubType: undefined,
}

HubTypeSelect.propTypes = {
  hubType: propTypes.string,
  setHubType: propTypes.func.isRequired,
  enableNext: propTypes.func.isRequired,
}

export default HubTypeSelect
