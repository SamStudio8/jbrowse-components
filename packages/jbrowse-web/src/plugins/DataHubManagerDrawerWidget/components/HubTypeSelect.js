import FormControl from '@material-ui/core/FormControl'
import FormControlLabel from '@material-ui/core/FormControlLabel'
import FormHelperText from '@material-ui/core/FormHelperText'
import Link from '@material-ui/core/Link'
import Radio from '@material-ui/core/Radio'
import RadioGroup from '@material-ui/core/RadioGroup'
import propTypes from 'prop-types'
import React from 'react'

const hubTypeDescriptions = {
  ucsc: (
    <FormHelperText>
      A track or assembly hub in the{' '}
      <Link
        href="http://genome.ucsc.edu/goldenPath/help/hgTrackHubHelp.html#Intro"
        rel="noopener noreferrer"
        target="_blank"
      >
        Track Hub
      </Link>{' '}
      format
    </FormHelperText>
  ),
  jbrowse1: (
    <FormHelperText>
      A{' '}
      <Link
        href="https://jbrowse.org/"
        rel="noopener noreferrer"
        target="_blank"
      >
        JBrowse 1
      </Link>{' '}
      data directory
    </FormHelperText>
  ),
}

function HubTypeSelect(props) {
  const { hubType, setHubType } = props
  const hubTypes = [
    { value: 'ucsc', label: 'Track or Assembly Hub' },
    { value: 'jbrowse1', label: 'JBrowse Hub' },
  ]
  return (
    <FormControl component="fieldset">
      <RadioGroup value={hubType} onChange={setHubType}>
        {hubTypes.map(entry => (
          <FormControlLabel
            key={entry.value}
            control={<Radio />}
            value={entry.value}
            label={entry.label}
            disabled={entry.value === 'jbrowse1'}
            data-testid={entry.value}
          />
        ))}
      </RadioGroup>
      {hubTypeDescriptions[hubType]}
    </FormControl>
  )
}

HubTypeSelect.defaultProps = {
  hubType: undefined,
}

HubTypeSelect.propTypes = {
  hubType: propTypes.string,
  setHubType: propTypes.func.isRequired,
}

export default HubTypeSelect
