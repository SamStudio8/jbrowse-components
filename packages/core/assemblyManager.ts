import AbortablePromiseCache from 'abortable-promise-cache'
import jsonStableStringify from 'json-stable-stringify'
import { autorun, observable } from 'mobx'
import {
  addDisposer,
  cast,
  getParent,
  IAnyType,
  SnapshotOrInstance,
  types,
} from 'mobx-state-tree'
import { readConfObject } from './configuration'
import { AnyConfigurationModel } from './configuration/configurationSchema'
import { Region } from './util/types'
import { Region as MSTRegion } from './util/types/mst'
import QuickLRU from './util/QuickLRU'

function checkRefName(refName: string) {
  // Valid refName pattern from https://samtools.github.io/hts-specs/SAMv1.pdf
  if (
    !refName.match(
      /[0-9A-Za-z!#$%&+./:;?@^_|~-][0-9A-Za-z!#$%&*+./:;=?@^_|~-]*/,
    )
  ) {
    throw new Error(`Encountered invalid refName: "${refName}"`)
  }
}

export function assemblyFactory(assemblyConfigType: IAnyType) {
  return types
    .model({
      configuration: types.reference(assemblyConfigType),
      regions: types.array(MSTRegion),
      refNameAliases: types.map(types.array(types.string)),
      refNameAliasesSet: false,
    })
    .views(self => ({
      get name(): string {
        return readConfObject(self.configuration, 'name')
      },
      get aliases(): string[] {
        return readConfObject(self.configuration, 'aliases')
      },
      get refNames() {
        return self.regions.map(region => region.refName)
      },
      get ready() {
        return !!self.regions.length && self.refNameAliasesSet
      },
      get allRefNames() {
        const aliases: string[] = []
        self.refNameAliases.forEach(aliasList => {
          aliases.push(...aliasList)
        })
        return [...this.refNames, ...aliases]
      },
      get rpcManager() {
        return getParent(self, 2).rpcManager
      },
      getCanonicalRefName(refName: string) {
        if (this.refNames.includes(refName)) {
          return refName
        }
        for (const [rName, aliases] of self.refNameAliases) {
          if (aliases.includes(refName)) {
            return rName
          }
        }
        return undefined
      },
      isValidRefName(refName: string) {
        return this.allRefNames.includes(refName)
      },
    }))
    .actions(self => ({
      afterAttach() {
        addDisposer(
          self,
          autorun(async () => {
            const sequenceAdapterConfig = readConfObject(self.configuration, [
              'sequence',
              'adapter',
            ])
            const adapterConfigId = jsonStableStringify(sequenceAdapterConfig)
            const sequenceAborter = new AbortController()
            const { rpcManager } = getParent(self, 2)
            rpcManager
              .call(
                adapterConfigId,
                'getRegions',
                {
                  sessionId: self.name,
                  adapterConfig: sequenceAdapterConfig,
                  signal: sequenceAborter.signal,
                },
                { timeout: 1000000 },
              )
              .then((adapterRegions: Region[]) => {
                const adapterRegionsWithAssembly = adapterRegions.map(
                  adapterRegion => {
                    const { refName } = adapterRegion
                    checkRefName(refName)
                    return {
                      ...adapterRegion,
                      assemblyName: self.name,
                    }
                  },
                )
                this.setRegions(adapterRegionsWithAssembly)
              })
              .catch((error: Error) => {
                getParent(self, 3).setError(String(error))
              })
            if (self.configuration.refNameAliases) {
              const refNameAliasesAdapterConfig = readConfObject(
                self.configuration,
                ['refNameAliases', 'adapter'],
              )
              const refNameAliasesAborter = new AbortController()
              rpcManager
                .call(
                  adapterConfigId,
                  'getRefNameAliases',
                  {
                    sessionId: self.name,
                    adapterConfig: refNameAliasesAdapterConfig,
                    signal: refNameAliasesAborter.signal,
                  },
                  { timeout: 1000000 },
                )
                .then(
                  (
                    refNameAliasesList: {
                      refName: string
                      aliases: string[]
                    }[],
                  ) => {
                    const refNameAliases: Record<string, string[]> = {}
                    refNameAliasesList.forEach(refNameAlias => {
                      refNameAlias.aliases.forEach(alias => {
                        checkRefName(alias)
                      })
                      refNameAliases[refNameAlias.refName] =
                        refNameAlias.aliases
                    })
                    this.setRefNameAliases(refNameAliases)
                    this.setRefNameAliasesSet(true)
                  },
                )
                .catch((error: unknown) => {
                  getParent(self, 3).setError(String(error))
                })
            } else {
              this.setRefNameAliasesSet(true)
            }
          }),
        )
      },
      setRegions(regions: Region[]) {
        self.regions = cast(regions)
      },
      setRefNameAliases(refNameAliases: Record<string, string[]>) {
        self.refNameAliases = cast(refNameAliases)
      },
      setRefNameAliasesSet(isSet: boolean) {
        self.refNameAliasesSet = isSet
      },
    }))
    .actions(self => {
      const refNameMapCache = new AbortablePromiseCache({
        // QuickLRU is a good backing cache to use, but you can use any
        // cache as long as it supports `get`, `set`, `delete`, and `keys`.
        cache: new QuickLRU({ maxSize: 1000 }),

        // the `fill` callback will be called for a cache miss
        async fill(
          { adapterConf }: { adapterConf: unknown },
          signal?: AbortSignal,
        ) {
          const refNameMap = observable.map({})
          if (!self.refNameAliasesSet) {
            return refNameMap
          }
          const stateGroupName = jsonStableStringify(adapterConf)

          const refNames = await self.rpcManager.call(
            stateGroupName,
            'getRefNames',
            {
              sessionId: self.name,
              adapterConfig: adapterConf,
              signal,
            },
            { timeout: 1000000 },
          )
          refNames.forEach((refName: string) => {
            checkRefName(refName)
            const aliases = self.refNameAliases.get(refName)
            if (aliases) {
              aliases.forEach(refNameAlias => {
                refNameMap.set(refNameAlias, refName)
              })
            } else {
              self.refNameAliases.forEach((configAliases, configRefName) => {
                if (configAliases.includes(refName)) {
                  refNameMap.set(configRefName, refName)
                  configAliases.forEach(refNameAlias => {
                    if (refNameAlias !== refName)
                      refNameMap.set(refNameAlias, refName)
                  })
                }
              })
            }
          })
          return refNameMap
        },
      })

      /**
       * get Map of `canonical-name -> adapter-specific-name`
       */
      async function getRefNameMapForAdapter(
        adapterConf: unknown,
        opts: { signal?: AbortSignal } = {},
      ) {
        const adapterConfigId = jsonStableStringify(adapterConf)
        return refNameMapCache.get(
          adapterConfigId,
          { adapterConf },
          opts.signal,
        )
      }

      /**
       * get Map of `adapter-specific-name -> canonical-name`
       */
      async function getReverseRefNameMapForAdapter(
        adapterConf: unknown,
        opts: { signal?: AbortSignal } = {},
      ) {
        const map = await getRefNameMapForAdapter(adapterConf, opts)
        const reversed = new Map()
        for (const [canonicalName, adapterName] of map) {
          reversed.set(adapterName, canonicalName)
        }
        return reversed
      }
      return { getRefNameMapForAdapter, getReverseRefNameMapForAdapter }
    })
}

export default function assemblyManagerFactory(assemblyConfigType: IAnyType) {
  return types
    .model({
      assemblies: types.array(assemblyFactory(assemblyConfigType)),
    })
    .views(self => ({
      get(assemblyName: string) {
        const canonicalName = this.aliasMap.get(assemblyName)
        return self.assemblies.find(
          assembly => assembly.name === (canonicalName || assemblyName),
        )
      },
      get aliasMap() {
        const aliases: Map<string, string> = new Map()
        self.assemblies.forEach(assembly => {
          if (assembly.aliases.length) {
            assembly.aliases.forEach(assemblyAlias => {
              aliases.set(assemblyAlias, assembly.name)
            })
          }
        })
        return aliases
      },
      get rpcManager() {
        return getParent(self).rpcManager
      },
      get allPossibleRefNames() {
        const refNames: string[] = []
        self.assemblies.forEach(assembly => {
          refNames.push(...assembly.refNames)
        })
        return refNames
      },
    }))
    .views(self => ({
      async getRefNameMapForAdapter(
        adapterConf: unknown,
        assemblyName: string,
        opts: { signal?: AbortSignal } = {},
      ) {
        const assembly = self.get(assemblyName)
        if (assembly) {
          return assembly.getRefNameMapForAdapter(adapterConf, opts)
        }
        return observable.map({})
      },
      async getReverseRefNameMapForAdapter(
        adapterConf: unknown,
        assemblyName: string,
        opts: { signal?: AbortSignal } = {},
      ) {
        const assembly = self.get(assemblyName)
        if (assembly) {
          return assembly.getReverseRefNameMapForAdapter(adapterConf, opts)
        }
        return observable.map({})
      },
      isValidRefName(refName: string, assemblyName?: string) {
        if (assemblyName) {
          const assembly = self.get(assemblyName)
          if (assembly) {
            return assembly.isValidRefName(refName)
          }
        }
        return self.allPossibleRefNames.includes(refName)
      },
    }))
    .actions(self => ({
      afterAttach() {
        // self.assemblies.push({
        //   configuration: getParent(self).jbrowse.assemblies[0],
        // })
        addDisposer(
          self,
          autorun(() => {
            getParent(self).jbrowse.assemblies.forEach(
              (assemblyConfig: AnyConfigurationModel) => {
                const existingAssemblyIdx = self.assemblies.findIndex(
                  assembly =>
                    assembly.name === readConfObject(assemblyConfig, 'name'),
                )
                if (existingAssemblyIdx !== -1) {
                  this.replaceAssembly(existingAssemblyIdx, assemblyConfig)
                } else {
                  this.addAssembly(assemblyConfig)
                }
              },
            )
          }),
        )
      },
      addAssembly(
        assemblyConfig: SnapshotOrInstance<typeof assemblyConfigType> | string,
      ) {
        self.assemblies.push({ configuration: assemblyConfig })
      },
      replaceAssembly(
        idx: number,
        assemblyConfig: SnapshotOrInstance<typeof assemblyConfigType> | string,
      ) {
        self.assemblies[idx] = assemblyConfigType.create({
          configuration: assemblyConfig,
        })
      },
    }))
}
