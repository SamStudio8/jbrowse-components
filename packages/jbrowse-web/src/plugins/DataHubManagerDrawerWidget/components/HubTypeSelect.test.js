import { createShallow } from '@material-ui/core/test-utils'
import React from 'react'
import HubTypeSelect from './HubTypeSelect'

describe('<HubTypeSelect />', () => {
  let shallow

  beforeAll(() => {
    shallow = createShallow()
  })

  it('renders', () => {
    const wrapper = shallow(
      <HubTypeSelect setHubType={() => {}} enableNext={() => {}} />,
    )
    expect(wrapper).toMatchSnapshot()
  })

  it('selects a hub type', () => {
    const setHubType = jest.fn(event => event.target.value)
    const enableNext = jest.fn(() => true)
    const wrapper = shallow(
      <HubTypeSelect setHubType={setHubType} enableNext={enableNext} />,
    )
    wrapper
      .find('RadioGroup')
      .props()
      .onChange({ target: { value: 'ucsc' } })
    expect(setHubType.mock.calls.length).toBe(1)
    expect(setHubType.mock.results[0].value).toBe('ucsc')
    expect(enableNext.mock.calls.length).toBe(1)
    expect(enableNext.mock.results[0].value).toBe(true)
  })
})
