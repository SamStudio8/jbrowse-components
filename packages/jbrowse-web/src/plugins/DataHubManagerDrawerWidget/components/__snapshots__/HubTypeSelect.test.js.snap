// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HubTypeSelect /> renders 1`] = `
<WithStyles(FormControl)
  component="fieldset"
>
  <RadioGroup
    onChange={[Function]}
  >
    <WithStyles(WithFormControlContext(FormControlLabel))
      control={<WithStyles(Radio) />}
      disabled={false}
      key="ucsc"
      label="Track or Assembly Hub"
      value="ucsc"
    />
    <WithStyles(WithFormControlContext(FormControlLabel))
      control={<WithStyles(Radio) />}
      disabled={true}
      key="jbrowse1"
      label="JBrowse Hub"
      value="jbrowse1"
    />
  </RadioGroup>
</WithStyles(FormControl)>
`;
